pulp_2to3_migration/app/models/base.py

```diff
@@ -11,24 +19,147 @@ class MigrationPlan(Model):
+        if self.plan_view.missing_repositories:
+            ret['repositories'] = self.plan_view.missing_repositories
+        if self.plan_view.missing_importers:
+            ret['importers'] = self.plan_view.missing_importers
+        if self.plan_view.missing_distributors:
+            ret['distributors'] = self.plan_view.missing_distributors
+        return ret
+
+
+class _InternalMigrationPlan:
+    def __init__(self, migration_plan):
+        self.migration_plan = migration_plan
+
+        self.plugins_to_migrate = []
+        self.importers_to_migrate = []
+        self.distributors_to_migrate = []
+        # pre-migration *just* needs these repos and nothing else
+        self.repositories_to_migrate = []
+
+        self.importer_discovery_repos = []
+
+        # a nested data structure with a format roughly matching the JSON schema.
+        # dictionary where the key is the name of the pulp3 repo and the value is a dict
+        # of other information like repo_versions, importer to use, etc.
+        self.repositories_to_create = {}
+
+        self.missing_importers = []
+        self.missing_repositories = []
+        self.missing_distributors = []
+
+        # Make sure we've initialized the MongoDB connection first
+        connection.initialize()
+        self._populate()
+        self._check_missing()
+
+    def _populate(self):
+        for plugin_data in self.migration_plan['plugins']:
+            self.plugins_to_migrate.append(plugin_data['type'])
+            if plugin_data.get('repositories'):
+                self._parse_repository_data(plugin_data.get('repositories'))
+                # TODO: do something with protection
+
+    def _parse_repository_data(self, repository_data):
+        for repository in repository_data:
+            name = repository['name']
+
+            _find_importer_repo = repository['pulp2_importer_repository_id']
+            self.importer_discovery_repos.append(_find_importer_repo)
+
+            repository_versions = self._parse_repository_version_data(
+                repository.get('repository_versions', [])
+            )
+
+            repository_data = {
+                "pulp2_importer_repository_id": _find_importer_repo,
+                "versions": repository_versions
+                # TODO: do something with protection
+            }
+
+            self.repositories_to_create[name] = repository_data
+
+    def _parse_repository_version_data(self, repository_version_data):
+        repository_versions = []
+
+        for repository_version in repository_version_data:
+            pulp2_repository_id = repository_version['pulp2_repository_id']
+            self.repositories_to_migrate.append(pulp2_repository_id)
+            repository_versions.append(pulp2_repository_id)
+
+            distributor_ids = repository_version.get('distributor_ids')
+            self.distributors_to_migrate.extend(distributor_ids)
+
+        return repository_versions
+
+    def _check_missing(self):
```

Comment:
Notably this does not modify the existing representations of repositories, importers, distributors to migrate.  Should it?